### PR TITLE
Feature/circular window

### PR DESCRIPTION
**Pull request by bkm**
* import the files of metagenlab_lib code into zdb (in webapp.lib, the zdb folder was not intended to contain code)
* add support for circular contigs in the genomic region viewer
* several bugfixes introduced in format_pathway and format_module

### CODE REVIEW

**Review by njohner — COMMENTED** [review]
LGTM, just a few remarks:
- How do you get these "merge master" commits? I usually just rebase, then the history is clean (although I'm pretty sure it was a mess in this case, seing how much stuff I changed and how old some of your commits were).
- I think you could squash some of your commits together. I started reviewing commit by commit, but you actually go over the same things in several separate commits, making it very hard to review. In the end I basically gave up.
- I quickly tested your branch (ran the tests), and two things came up:
  - We still need https://github.com/metagenlab/zDB/blob/master/webapp/manage.py#L5C1-L5C77 for testing purposes
  - We need https://github.com/metagenlab/metagenlab_libs/pull/7 in order for things to work.
  - You need to remove that line: https://github.com/metagenlab/zDB/blob/master/testing/webapp/test_views.py#L10

I've made these changes locally and could commit and push them to your branch if you want.

Also what is the plan on the metagenlab_libs side? Are we deleting from that repo the code we moved over?

File: webapp/chlamdb/views.py [hunk 1]
```diff
@@ -1676,17 +1676,37 @@ def locusx_genomic_region(db, seqid, window):
...
+        diff = contig_size+window_start
+        df_seqids = df_seqids[(df_seqids.end_pos >= diff) | (df_seqids.start_pos <= window_stop)]
+        df_seqids.start_pos += -window_start
+        df_seqids.end_pos += -window_start
+        df_seqids.start_pos -= (df_seqids.start_pos>contig_size)*contig_size
+        df_seqids.end_pos -= (df_seqids.end_pos>contig_size)*contig_size
+        window_stop += -window_start
         window_start = 0
-        window_stop = 0
+        df_seqids.start_pos -= (df_seqids.start_pos>window_stop)*df_seqids.start_pos
```
[the remark this answers is not in the document]
Comment (bkm) [reply webapp/chlamdb/views.py]: Actually, it's also an outdated version. I think the current one is clearer as it avoids the multiplication with contig size.

File: webapp/chlamdb/views.py [hunk 2]
```diff
@@ -1676,17 +1676,37 @@ def locusx_genomic_region(db, seqid, window):
...
     elif window_stop>contig_size and not is_circular:
         window_stop = contig_size
-        df_seqids = df_seqids[stop <= contig_size]
+        df_seqids = df_seqids[df_seqids.end_pos>window_start]
     elif window_start<0:
+        # circular contig
+
+        diff = contig_size+window_start
+        df_seqids = df_seqids[(df_seqids.end_pos >= diff) | (df_seqids.start_pos <= window_stop)]
+        df_seqids.start_pos += -window_start
```
[the remark this answers is not in the document]
Comment (bkm) [reply webapp/chlamdb/views.py]: Changed.

**Comment by bkm** [comment]
I think I adressed most issues. Maybe have a look at the final diff : some issues you raised were already solved. Thanks for the remark about PEP8, I'll find a vim plugin.

I used the git merge command. And indeed, the commits were a bit hard to follow, sorry about that. I need to get used to working with several people on the same project again.

About metagenlab_libs, we need to check with Trestan. As for now, the repo is not needed for zdb anymore, so we could indeed delete the legacy code.

As for the warnings, I did not have time to look into them yet. I'll do that the upcoming days.

**Comment by njohner** [comment]
Yes I saw that some remarks were for outdated code. I got used to reviewing on a per commit basis, but that only makes sense on a clean history. I didn't really want to look at the whole diff, because of the merge commits and because of moving the metagenlab_libs over.
So anyway, feel free to merge, as it's a rather big PR, merging it rapidly will avoid too much rebase mess...
